File: examples/someipy_demo/client_fusion.cpp
#include <fusion_hawking/runtime.hpp>
#include <iostream>
#include <thread>
#include <chrono>

#ifdef _WIN32
#include <ws2tcpip.h>
#endif
#ifndef _WIN32
#include <arpa/inet.h>
#include <unistd.h>
#include <netinet/in.h>
#endif

using namespace fusion_hawking;

int main() {
    try {
        SomeIpRuntime runtime("client_config.json", "cpp_client");
        // runtime starts in constructor via jthread reactor

        std::cout << "[Fusion C++ Client] Waiting for Service 0x1234..." << std::endl;
        
        if (runtime.wait_for_service(0x1234, 0x0001)) {
            sockaddr_storage remote_ep;
            if (runtime.get_remote_service(0x1234, 0x0001, remote_ep)) {
                
                char ip_str[INET6_ADDRSTRLEN];
                uint16_t port = 0;
                if (remote_ep.ss_family == AF_INET) {
                    inet_ntop(AF_INET, &((sockaddr_in*)&remote_ep)->sin_addr, ip_str, INET_ADDRSTRLEN);
                    port = ntohs(((sockaddr_in*)&remote_ep)->sin_port);
                } else {
                    inet_ntop(AF_INET6, &((sockaddr_in6*)&remote_ep)->sin6_addr, ip_str, INET6_ADDRSTRLEN);
                    port = ntohs(((sockaddr_in6*)&remote_ep)->sin6_port);
                }

                std::cout << "[Fusion C++ Client] Discovered service at " << ip_str << ":" << port << std::endl;
                
                std::string msg = "Hello from Fusion C++!";
                std::vector<uint8_t> payload(msg.begin(), msg.end());
                
                auto response = runtime.SendRequest(0x1234, 0x0001, payload, remote_ep);
                
                if (!response.empty()) {
                    std::string res_str(response.begin(), response.end());
                    std::cout << "[Fusion C++ Client] Got Response: '" << res_str << "'" << std::endl;
                } else {
                    std::cout << "[Fusion C++ Client] RPC Timeout or Error" << std::endl;
                }
            }
        } else {
            std::cout << "[Fusion C++ Client] Service not found (Timeout)." << std::endl;
        }

        // Destructor handles stop
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}

File: examples/someipy_demo/client_fusion.rs
use fusion_hawking::runtime::{SomeIpRuntime, ServiceClient};
use fusion_hawking::transport::SomeIpTransport;
use std::net::SocketAddr;
use std::sync::Arc;
use std::thread;
use std::time::Duration;

#[derive(Clone)]
struct GenericClient {
    #[allow(dead_code)]
    transport: Arc<dyn SomeIpTransport>,
    target: SocketAddr,
}

// ServiceClient usually requires Send + Sync if it's stored in the runtime or used across threads
impl ServiceClient for GenericClient {
    const SERVICE_ID: u16 = 0x1234;
    fn new(transport: Arc<dyn SomeIpTransport>, target: SocketAddr) -> Self {
        Self { transport, target }
    }
}

// Explicitly implement Send and Sync if compiler is unsure, 
// though Arc<dyn SomeIpTransport> should already be Send+Sync if the trait is.
unsafe impl Send for GenericClient {}
unsafe impl Sync for GenericClient {}

#[tokio::main]
async fn main() {
    let args: Vec<String> = std::env::args().collect();
    let default_path = "client_config.json".to_string();
    let config_path = if args.len() > 1 { &args[1] } else { &default_path };
    let instance_name = "rust_client";

    println!("[Fusion Rust Client] Initializing with config: {}", config_path);
    let runtime = SomeIpRuntime::load(config_path, instance_name);
    
    let rt_clone = runtime.clone();
    thread::spawn(move || {
        rt_clone.run();
    });

    println!("[Fusion Rust Client] Searching for Service 0x1234...");
    thread::sleep(Duration::from_secs(1));

    if let Some(client) = runtime.get_client::<GenericClient>("someipy_svc") {
        println!("[Fusion Rust Client] Discovered service at {}", client.target);
        
        let msg = "Hello from Fusion Rust!";
        let payload = msg.as_bytes().to_vec();
        
        println!("[Fusion Rust Client] Sending Echo: '{}'", msg);
        match runtime.send_request_and_wait(0x1234, 0x0001, &payload, client.target).await {
            Some(response) => {
                let res_str = String::from_utf8_lossy(&response);
                println!("[Fusion Rust Client] Got Response: '{}'", res_str);
            }
            None => {
                println!("[Fusion Rust Client] RPC Error: No response received");
            }
        }
    } else {
        println!("[Fusion Rust Client] Service not found.");
    }

    runtime.stop();
    thread::sleep(Duration::from_millis(500));
}
